### ISSUE TITLE
Feature idea: parameter files

### ISSUE THREAD

**Issue by goonzoid** [CrushedPixel/Plugalyzer#3]
Continuing our conversation from discord... 😃 

Setting lots of parameters on the command line can get a little unwieldy, so it might be useful to be able to provide all parameter settings via a single file, probably json formatted.

There could be a `--paramFile=<path>` option. If `--param=<name>:<value>` options are also provided, they could override what's in the file, or it might be simpler to say that it's invalid to use both flags together. Not sure what's best here.

Your idea to have the `listParameters` command also output the parameter list in json format (maybe optionally via another flag?) would also be useful in this context.

As discussed, I'd be happy to make a PR for this in a few weeks time.

**Comment by CrushedPixel** [comment]
I think `--params=<path>` would be my preferred argument name for the input file. additional `--param` options should override the one set in the file, with a note to `std::cout` making the user aware of the conflict.

The easiest way to make `listParameters` output a JSON file is by simply adding the `--output=<path>` parameter. We can still print the parameters to `std::cout` as we currently do.

**Comment by goonzoid** [comment]
Just thinking this through a bit more, there's possibly a difference between the json that might be expected from the output of `listParameters` vs the json that would be useful as an input to `process`.

e.g. for `process`, I might just want to supply something like:
```json
[
    {"name": "Power", "value": "On"},
    {"name": "Saturation Frequency", "value": "Flat"},
    {"name": "Saturation In", "value": "Off"},
    {"name": "Saturation", "value": "50%"}
]
```

whereas for the output of `listParameters` I would probably expect to see something like:
```json
{
    "name":"Black Box Analog Design HG-2",
    "parameters": [
        {
            "index": 0,
            "name": "Power",
            "discrete_values": ["Off", "On"],
            "default_value": "On"
        },
        {
            "index": 1,
            "name": "Saturation Frequency",
            "discrete_values": ["Low", "Flat", "High"],
            "default_value": "Flat"
        },
        {
            "index": 2,
            "name": "Saturation In",
            "discrete_values": ["Off", "On"],
            "default_value": "Off"
        },
        {
            "index": 3,
            "name": "Saturation",
            "value_range": {"min": "0%", "max": "100%"},
            "default_value": "50%"
        }
    ]
}
```

So, it depends a bit on what we think the use case for outputting json from the `listParameters` command is. If it's just to make it easier to supply params to the `process` flag, then we could output something simpler like the first example above (at which point, maybe it isn't even worth it?).

What do you think? I'm open to any suggestions.

**Comment by CrushedPixel** [comment]
I'd simplify the JSON input for `process` even further:
```json
{
    "Power": "On",
    "Saturation Frequency": "Flat",
    "Saturation In": "Off",
    "Saturation": "50%"
}
```

This makes it a bit easier to work with, and since parameter names should be unique, we can use them as keys.

I can't really think of a use-case for the JSON list of parameters right now - it'd be useful when either human intervention in parameter setting isn't required, or when visualizing the parameters in a tool that doesn't want to load the plugin itself for some reason. 
I like the format you're proposing for the JSON output, except that I prefer camelCase over snake_case for the JSON keys - it usually maps better to field names when you parse them into objects in your code.

**Comment by sudara** [comment]
Just want to say I'm happy you cooked up this tool!

> I can't really think of a use-case for the JSON list of parameters right now

I'd have a use case (assuming midi support became a thing, since I'm mainly building a synth) — I'd like to run phase cancellation tests in CI against a set of reference WAV files. Ideally each test would be 2 pieces of data, an input and a list of parameters.

**Comment by CrushedPixel** [comment]
I'm definitely willing to add support for automation. I've implemented multiple keyframe and interpolation systems before, so it would probably be quite trivial.

How about the following syntax/spec for automation in a JSON file:
```json5
{
    "Power": "On",   // no automation

    // keyframe locations specified in samples:
    "Dry/Wet Mix": { 
        // keyframe at 0 is not required, 
        // it will use the value of the first keyframe until that keyframe is reached
        "13024": "50%",
        "22050": "70%",
        "52316": "20%"
    },

    // keyframe locations specified in seconds:
    "Saturation": {  
        "0s":     "20%", 
        "2.012s": "40%",
        "6.5s":   "100%"
    },

    // keyframe locations specified in percentage of total input duration:
    "Detune": {
        "0%":   "12",
        "50%":  "-12",
        "100%": "12"
    },

    // it's possible to mix keyframe time units:
    "Fuzziness": {
        "0":   "15%",
        "50%": "0%",
        "7s":  "100%"
    },

    // special parameter to specify interpolation type:
    "Color": {
        "interpolation": "cubic", // linear | cubic | catmull-rom
        "0":   "15%",
        "50%": "0%",
        "7s":  "100%"
    },
}
```

**Comment by sudara** [comment]
Wow. Automation would really turn this into a crazy useful tool for CI.

I'll respond based on my current use case (running midi files through a synth, getting audio output files, checking against reference audio files in CI).

* With no keyframe at 0/0%, I wonder if it's best to use the parameter's default value vs. the next value in the list. Perhaps that would be too implicit. 
* +1 to using the actual parameter ID as a key, as parameters are unique and that's what I'm using in my preset json, etc
* Initially I was imagining just using "seconds", but samples might open up the ability to test some detail/corner cases. Percentages would actually be pretty useful for highest level ("just ramp up this value over the duration")
* Interpolation is not a "must have" for me, as my focus is tests and that probably gets out of scope. Switching interpolations is def not in scope for me, I can't think of a use case.

Basically: looks amazing to have any variety of automation!

**Comment by CrushedPixel** [comment]
> With no keyframe at 0/0%, I wonder if it's best to use the parameter's default value vs. the next value in the list.

In my experience using and creating keyframe systems, this would be extremely unexpected behaviour. I'm firm on this one, we should go for the next value in the list.

> Interpolation is not a "must have" for me, as my focus is tests and that probably gets out of scope. Switching interpolations is def not in scope for me, I can't think of a use case.

Let's remove different interpolation types from the goals of this feature, then.

**Comment by CrushedPixel** [comment]
Here we go - you can now supply a parameter file that supports automation using `--paramFile`. For more information, refer to the README.
